README
Gnome2::Panel version 0.02
==========================

This module allows you to create applets for the GNOME 2.0 Panel using Perl.

INSTALLATION
------------

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES
------------

This module requires these other modules and libraries:

  perl                >= 5.8.0
  ExtUtils::Depends   >= 0.20  (Perl module)
  ExtUtils::PkgConfig >= 1.03  (Perl module)
  Glib                >= 1.160 (Perl module and requisite C libraries)
  Gtk2                >= 1.040 (Perl module and requisite C libraries)
  Gnome2              >= 1.042 (Perl module and requisite C libraries)
  Gnome2::GConf       >= 1.044 (Perl module and requisite C libraries)
  libpanelapplet-2.0  >= 2.0   (C library)


COPYRIGHT AND LICENSE
---------------------

Copyright (C) 2003, 2007 by the gtk2-perl team

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option) any
later version.

This library is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public License along
with this library; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA


NEWS
Overview of changes in Gnome2::PanelApplet 0.02
===============================================

* Wrap the GConf helper API.

Overview of changes in Gnome2::PanelApplet 0.01
===============================================

* Initial release.


libpanelapplet-perl.h
/*
 * Copyright (C) 2007 by the gtk2-perl team
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id$
 */

#ifndef _LIBPANELAPPLET_PERL_H_
#define _LIBPANELAPPLET_PERL_H_

#include <gnome2perl.h>

#include <panel-applet.h>
#include <panel-applet-enums.h>

/* Custom GType for the panel orientations. */
#define PANEL_PERL_TYPE_PANEL_APPLET_ORIENT panel_perl_applet_orient_get_type()
GType panel_perl_applet_orient_get_type (void);

#include "libpanelapplet-perl-autogen.h"
#include "libpanelapplet-perl-version.h"

#endif /* _LIBPANELAPPLET_PERL_H_ */


lib/Gnome2/PanelApplet.pm
package Gnome2::PanelApplet;

# $Id$

use 5.008;
use strict;
use warnings;

use Gnome2;
use Gnome2::GConf;

require DynaLoader;

our @ISA = qw(DynaLoader);

our $VERSION = '0.02';

sub import {
  my $self = shift();
  $self -> VERSION(@_);
}

sub dl_load_flags { 0x01 }

Gnome2::PanelApplet -> bootstrap($VERSION);

sub gconf_get_list {
  my ($applet, $key, $check_error) = @_;
  $check_error = 1 unless defined $check_error;

  my $value = $applet->gconf_get_value ($key, $check_error);
  return $value->{value};
}

sub gconf_set_list {
  my ($applet, $key, $list_type, $list, $check_error) = @_;
  $check_error = 1 unless defined $check_error;

  $applet->gconf_set_value ($key,
                            { type => $list_type,
                              value => $list },
                            $check_error);
}

1;
__END__

=head1 NAME

Gnome2::PanelApplet - Perl interface to GNOME's applet library

=head1 SYNOPSIS

  # Initialize.
  Gnome2::Program->init ('My Applet', '0.01', 'libgnomeui'
                         sm_connect => FALSE);

  # Register our applet with that bonobo thingy.  The OAFIID stuff is
  # specified in a .server file.  See
  # C<examples/GNOME_PerlAppletSample.server> in the
  # I<Gnome2::PanelApplet> tarball for an example.
  Gnome2::PanelApplet::Factory->main (
    'OAFIID:PerlSampleApplet_Factory', # iid of the applet
    'Gnome2::PanelApplet',             # type of the applet
    \&fill                             # sub that populates the applet
  );

  sub fill {
    my ($applet, $iid, $data) = @_;

    # Safety measure: if we're passed the wrong IID, just return.
    if ($iid ne 'OAFIID:PerlSampleApplet') {
      return FALSE;
    }

    # Gnome2::PanelApplet isa Gtk2::EventBox, so it isa Gtk2::Container
    # in particular.  That means we can call add() on it.
    my $label = Gtk2::Label->new ('Hi, there!');
    $applet->add ($label);
    $applet->show_all;

    return TRUE;
  }

=head1 ABSTRACT

Use Perl to write GNOME applets that sit on the panel.

=head1 DOCUMENTATION

I<Gnome2::PanelApplet::Factory-E<gt>main> is documented in
L<Gnome2::PanelApplet::Factory>.  The methods you can call on the applet
instance are documented in L<Gnome2::PanelApplet::main>.  The GConf helper
functions are documented in L<Gnome2::PanelApplet::GConf>.

=head1 SEE ALSO

L<Gnome2::PanelApplet::index>, L<Gnome2>, L<Gtk2>, L<Gtk2::api> and
L<http://developer.gnome.org/doc/API/2.0/panel-applet/>.

=head1 AUTHOR

Emmanuele Bassi E<lt>emmanuele.bassi at iol dot itE<gt>

Torsten Schoenfeld E<lt>kaffeetisch at gmx dot deE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003, 2007 by the gtk2-perl team

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option) any
later version.

This library is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public License along
with this library; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

=cut


Makefile.PL
#!/usr/bin/perl
# $Id$

use 5.008;
use strict;
use warnings;

use Cwd qw(cwd);
use ExtUtils::MakeMaker;

our %build_reqs = (
  'libpanelapplet-2.0' => '2.0.0',
);

our %perl_reqs = (
  'ExtUtils::Depends'   => '0.20',
  'ExtUtils::PkgConfig' => '1.03',
  'Glib'                => '1.160',
  'Gtk2'                => '1.040',
  'Gnome2'              => '1.042',
  'Gnome2::GConf'       => '1.044',
);

unless (eval "use ExtUtils::Depends '$perl_reqs{'ExtUtils::Depends'}';"
           . "use ExtUtils::PkgConfig '$perl_reqs{'ExtUtils::PkgConfig'}';"
           . "use Glib '$perl_reqs{'Glib'}';"
           . "use Gtk2 '$perl_reqs{'Gtk2'}';"
           . "use Gnome2 '$perl_reqs{'Gnome2'}';"
           . "use Gnome2::GConf '$perl_reqs{'Gnome2::GConf'}';"
           . "use Glib::MakeHelper;"
           . "use Gtk2::CodeGen;"
           . "1") {
  warn "$@\n";
  WriteMakefile(
    PREREQ_FATAL => 1,
    PREREQ_PM    => \%perl_reqs,
  );
  exit 1; # not reached
}

my %pkgcfg = ExtUtils::PkgConfig->find("libpanelapplet-2.0 >= $build_reqs{'libpanelapplet-2.0'}");

mkdir 'build', 0777;

our @xs_files = <xs/*.xs>;
our %pod_files = (
  'lib/Gnome2/PanelApplet.pm' => '$(INST_MAN3DIR)/Gnome2::PanelApplet.$(MAN3EXT)',
  Glib::MakeHelper->do_pod_files (@xs_files),
);

Gtk2::CodeGen->parse_maps('libpanelapplet-perl');
Gtk2::CodeGen->write_boot(ignore => qr/^Gnome2::PanelApplet$/);

ExtUtils::PkgConfig->write_version_macros (
  "build/libpanelapplet-perl-version.h",
  "libpanelapplet-2.0" => "PANEL_APPLET",
);

my $dep = ExtUtils::Depends->new('Gnome2::PanelApplet', qw/Gnome2 Gnome2::GConf/);
$dep->set_inc($pkgcfg{cflags});
$dep->set_libs($pkgcfg{libs});
$dep->add_xs(@xs_files);
$dep->add_pm('lib/Gnome2/PanelApplet.pm' => '$(INST_LIBDIR)/PanelApplet.pm');
$dep->add_typemaps(map {File::Spec->catfile(cwd(), $_)} 'build/libpanelapplet-perl.typemap');

$dep->install(qw(libpanelapplet-perl.h build/libpanelapplet-perl-autogen.h build/libpanelapplet-perl-version.h));
$dep->save_config('build/IFiles.pm');

WriteMakefile(
  NAME          => 'Gnome2::PanelApplet',
  VERSION_FROM  => 'lib/Gnome2/PanelApplet.pm',
  ABSTRACT_FROM => 'lib/Gnome2/PanelApplet.pm',
  XSPROTOARG    => '-noprototypes',
  MAN3PODS      => \%pod_files,
  PREREQ_PM     => \%perl_reqs,
  $dep->get_makefile_vars,
);

sub MY::postamble {
	return Glib::MakeHelper->postamble_clean ()
	     . Glib::MakeHelper->postamble_docs_full (
		DEPENDS => $dep,
		COPYRIGHT_FROM => 'copyright.pod');
}
